MSAL/src/MSALPublicClientApplication.m

```diff
@@ -1332,6 +1322,56 @@ - (void)signoutWithAccount:(nonnull MSALAccount *)account
     }];
 }
 
+#pragma mark - Device information
+
+- (void)getDeviceInformationWithParameters:(MSALParameters *)parameters
+                           completionBlock:(MSALDeviceInformationCompletionBlock)completionBlock
+{
+    MSID_LOG_WITH_CTX_PII(MSIDLogLevelInfo, nil, @"Querying device info");
+    
+    __auto_type block = ^(MSALDeviceInformation * _Nullable deviceInformation, NSError * _Nullable msidError)
+    {
+        NSError *msalError = nil;
+        
+        if (msidError)
+        {
+            msalError = [MSALErrorConverter msalErrorFromMsidError:msidError classifyErrors:YES msalOauth2Provider:self.msalOauth2Provider];
+        }
+        else
+        {
+            MSID_LOG_WITH_CTX_PII(MSIDLogLevelInfo, nil, @"Retrieved device info %@", deviceInformation);
+        }
+        
+        [MSALPublicClientApplication logOperation:@"getDeviceInformation" result:nil error:msalError context:nil];
+        
+        if (!completionBlock) return;
+        
+        if (parameters.completionBlockQueue)
+        {
+            dispatch_async(parameters.completionBlockQueue, ^{
+                completionBlock(deviceInformation, msalError);
+            });
+        }
+        else
+        {
+            completionBlock(deviceInformation, msalError);
+        }
+    };
+    
+    MSALDeviceInfoProvider *request = [MSALDeviceInfoProvider new];
+    
+    NSError *requestParamsError;
+    MSIDRequestParameters *requestParams = [self defaultRequestParametersWithError:&requestParamsError];
+    
+    if (!requestParams)
+    {
+        block(nil, requestParamsError);
+        return;
+    }
+    
+    [request deviceInfoWithRequestParameters:requestParams completionBlock:block];
```

Comment:
`request` is a local variable, can it be deallocated before we get a result in completion block?

Reply:
It captures self internally, so won't be deallocated. 
I was actually thinking to make it a property on public client app, so we can have one device info request at a time for one public client app. However, I think we can implement it in many different ways, so let me know if you have any recommendations here.